File: frontend/src/api/services/teamService.ts
import api from '../axios';
import type { TeamProfile, JoinRequest, TeamRoster } from '../../types';

// Types for API requests/responses
export interface CreateTeamRequest {
    teamName: string;
    description?: string;
    logoUrl?: string;
    location?: string;
    latitude?: number;
    longitude?: number;
    skillLevel?: number;
}

export interface UpdateTeamRequest extends Partial<CreateTeamRequest> {
    teamId: number;
}

export interface TeamSearchParams {
    query?: string;
    location?: string;
    skillLevel?: number;
    status?: string;
    page?: number;
    limit?: number;
}

export interface PaginatedResponse<T> {
    data: T[];
    total: number;
    page: number;
    limit: number;
    totalPages: number;
}

// Team service functions - Real API calls
export const teamService = {
    /**
     * Create a new team
     */
    createTeam: async (data: CreateTeamRequest): Promise<TeamProfile> => {
        const response = await api.post<TeamProfile>('/teams', data);
        return response.data;
    },

    /**
     * Get team by ID
     */
    getTeamById: async (teamId: number): Promise<TeamProfile> => {
        const response = await api.get<TeamProfile>(`/teams/${teamId}`);
        return response.data;
    },

    /**
     * Update team profile
     */
    updateTeam: async ({ teamId, ...data }: UpdateTeamRequest): Promise<TeamProfile> => {
        const response = await api.put<TeamProfile>(`/teams/${teamId}`, data);
        return response.data;
    },

    /**
     * Delete team
     */
    deleteTeam: async (teamId: number): Promise<void> => {
        await api.delete(`/teams/${teamId}`);
    },

    /**
     * Get teams led by current user
     */
    getLeaderTeams: async (_userId: number): Promise<TeamProfile[]> => {
        const response = await api.get<TeamProfile[]>('/teams/my-teams');
        return response.data;
    },

    /**
     * Get teams where player is a member
     */
    getPlayerTeams: async (playerId: number): Promise<TeamProfile[]> => {
        const response = await api.get<TeamProfile[]>(`/teams/player/${playerId}`);
        return response.data;
    },

    /**
     * Search teams
     */
    searchTeams: async (params: TeamSearchParams): Promise<PaginatedResponse<TeamProfile>> => {
        const response = await api.get<TeamProfile[]>('/search/teams', { params });
        // Wrap in paginated response format
        return {
            data: response.data,
            total: response.data.length,
            page: params.page || 1,
            limit: params.limit || 20,
            totalPages: 1,
        };
    },

    /**
     * Request to join a team
     */
    requestJoinTeam: async (teamId: number, message?: string): Promise<JoinRequest> => {
        const response = await api.post<JoinRequest>(`/teams/${teamId}/join-requests`, { message });
        return response.data;
    },

    /**
     * Leave a team
     */
    leaveTeam: async (teamId: number, playerId: number): Promise<void> => {
        await api.delete(`/teams/${teamId}/roster/${playerId}`);
    },

    /**
     * Get pending join requests for a team
     */
    getPendingJoinRequests: async (teamId: number): Promise<JoinRequest[]> => {
        const response = await api.get<JoinRequest[]>(`/teams/${teamId}/join-requests`);
        return response.data;
    },

    /**
     * Process join request (approve/reject)
     */
    processJoinRequest: async (requestId: number, approve: boolean): Promise<JoinRequest> => {
        const action = approve ? 'accept' : 'reject';
        const response = await api.put<JoinRequest>(`/teams/join-requests/${requestId}/${action}`);
        return response.data;
    },

    /**
     * Get team roster
     */
    getTeamRoster: async (teamId: number): Promise<TeamRoster[]> => {
        const response = await api.get<TeamRoster[]>(`/teams/${teamId}/roster`);
        return response.data;
    },
};

export default teamService;

File: backend/app/services/team_service.py
"""
TeamService - Team management business logic.
Maps to TeamController in class diagram.
"""
from typing import List, Optional
from datetime import datetime
from sqlalchemy.ext.asyncio import AsyncSession

from app.repositories.team_repository import TeamRepository, RosterRepository, JoinRequestRepository
from app.repositories.player_repository import PlayerRepository
from app.models.team import TeamProfile, TeamRoster, JoinRequest
from app.models.enums import TeamStatus, RosterRole, JoinRequestStatus, UserRole


class TeamService:
    """Service handling team business logic."""
    
    def __init__(self, db: AsyncSession):
        self.db = db
        self.team_repo = TeamRepository(db)
        self.roster_repo = RosterRepository(db)
        self.join_request_repo = JoinRequestRepository(db)
        self.player_repo = PlayerRepository(db)
    
    async def create_team(
        self,
        leader_id: int,
        team_name: str,
        description: str = None,
        location: str = None,
        latitude: float = None,
        longitude: float = None,
    ) -> TeamProfile:
        """Create a new team with the user as leader."""
        # Check name uniqueness
        existing = await self.team_repo.find_by_name(team_name)
        if existing:
            raise ValueError("Team name already exists")
        
        team = TeamProfile(
            team_name=team_name,
            description=description,
            leader_id=leader_id,
            status=TeamStatus.PENDING,
            location=location,
            latitude=latitude,
            longitude=longitude,
        )
        await self.team_repo.save(team)
        
        # Add leader to roster as Captain
        player = await self.player_repo.find_by_user_id(leader_id)
        if player:
            roster_entry = TeamRoster(
                team_id=team.team_id,
                player_id=player.player_id,
                role=RosterRole.CAPTAIN,
                is_active=True,
            )
            await self.roster_repo.save(roster_entry)
        
        await self.team_repo.commit()
        return team
    
    async def get_team_by_id(self, team_id: int) -> Optional[TeamProfile]:
        """Get team by ID."""
        return await self.team_repo.find_by_id(team_id)
    
    async def get_teams_by_leader(self, leader_id: int) -> List[TeamProfile]:
        """Get teams where user is leader."""
        return await self.team_repo.find_by_leader_id(leader_id)
    
    async def update_team(
        self,
        team: TeamProfile,
        team_name: str = None,
        description: str = None,
        **kwargs
    ) -> TeamProfile:
        """Update team profile."""
        if team_name and team_name != team.team_name:
            existing = await self.team_repo.find_by_name(team_name)
            if existing:
                raise ValueError("Team name already exists")
            team.team_name = team_name
        
        if description is not None:
            team.description = description
        
        for key, value in kwargs.items():
            if hasattr(team, key) and value is not None:
                setattr(team, key, value)
        
        await self.team_repo.update(team)
        await self.team_repo.commit()
        return team
    
    async def delete_team(self, team: TeamProfile) -> bool:
        """Delete a team."""
        await self.team_repo.delete(team)
        await self.team_repo.commit()
        return True
    
    # --- Join Requests ---
    
    async def create_join_request(
        self,
        team_id: int,
        player_id: int,
        message: str = None
    ) -> JoinRequest:
        """Create a join request."""
        # Check for existing pending request
        existing = await self.join_request_repo.find_pending_by_player(player_id, team_id)
        if existing:
            raise ValueError("You already have a pending request")
        
        join_request = JoinRequest(
            team_id=team_id,
            player_id=player_id,
            status=JoinRequestStatus.PENDING,
            message=message,
        )
        await self.join_request_repo.save(join_request)
        await self.join_request_repo.commit()
        return join_request
    
    async def get_pending_requests(self, team_id: int) -> List[JoinRequest]:
        """Get pending join requests for a team."""
        return await self.join_request_repo.find_pending_by_team(team_id)
    
    async def process_join_request(
        self,
        request: JoinRequest,
        approve: bool
    ) -> bool:
        """Approve or reject a join request."""
        request.processed_at = datetime.utcnow()
        
        if approve:
            request.status = JoinRequestStatus.ACCEPTED
            # Add player to roster
            roster_entry = TeamRoster(
                team_id=request.team_id,
                player_id=request.player_id,
                role=RosterRole.MEMBER,
                is_active=True,
            )
            await self.roster_repo.save(roster_entry)
        else:
            request.status = JoinRequestStatus.REJECTED
        
        await self.join_request_repo.update(request)
        await self.join_request_repo.commit()
        return True
